Fix reloading page issue when user make Stake S3F/ Unstake S3F actions on page “Earn SNOB”  (deprecated)

thongpt-codelink opened the pull request:
## Status
**Done**

## Description
- The logic for fetching information and rendering the poolS3F is now separated, no need to wait until other fetching resolved.
- Instead of reloading the page when the action is submitted, fetching only relevant information and re-rendering the element that the action interact.

https://user-images.githubusercontent.com/83624863/117252489-38f3bd00-ae70-11eb-9d71-bd6709562fa4.mov

thongpt-codelink committed:
Change the logic for rendering poolS3F element
- trim() the pool_name and use it as element id
- if the id exist, replace the old element, if not, create new element

thongpt-codelink committed:
Add new method for getting relevant information and rendering poolS3F element
- move poolS3F function definition to the global scope
- create renderPoolS3F function for getting relevant info then render poolS3F

thongpt-codelink committed:
Code review
- change trim() to split(' ').join('') to get poolId, which is more appropriate
- move stakeUnstake definition to global scope, before it is in main()
- change all var to let in poolS3F

thongpt-codelink committed:
Re-render the poolS3F separatedly instead of reload page when staking S3F
- change stakingContract_stake arguments, to object for clearance, add STAKING_CONTRACT, SNOB_TOKEN, S3F_TOKEN, renderPoolS3F for re-rendering the element
- change stakingContract_withdraw arguments, to object for clearance, add STAKING_CONTRACT, SNOB_TOKEN, S3F_TOKEN, renderPoolS3F for re-rendering the element
- add snobPrice to poolS3F arguments
- call function renderPoolS3F instead of window.reload() when the process succeded
- move renderPoolS3F to the main() scope for stakingContract_stake execution

thongpt-codelink committed:
Fix the new rendered element is not binded to the data-attr set before
- re-invoke the jQuery function to bind handler to new created element
- move stakingContract_stake, stakingContract_withdraw to inside the main() scope so the invoked renderPoolS3F can reuse the jQuery method for data-btn (the handler is declared in the main() scope, this prevents additional change for the current codebase)

thongpt-codelink committed:
Move renderPoolS3F function call to the old position of poolS3F
- this approach (from the last development) cause the slow issue of the UI, cause the UI only rendered after all Promises to retrieve data is resolved

thongpt-codelink committed:
Remove commented out code
- remove the unused snippet of stakingContract_stake, stakingContract_withdraw
- remove the old poolS3F function call, replace with the new renderPoolS3F

Commit:
Fix missing Harvest SNOB button
- change fetching handler for pendingSNOBTokensPool8 in renderPoolS3F to STAKING_CONTRACT.earned(), use wrong handler before

Commit:
Code review: use jQuery.unbind() to remove the registered handler before add that handler to prevent duplicated execution

huynvk commented on src/static/js/earn.js:
```diff
@@ -73,6 +74,88 @@ async function main() {
+    STAKING_ABI, STAKING_ADDR, S3F_ADDRESS, App, STAKING_CONTRACT, SNOB_TOKEN, S3F_TOKEN, renderPoolS3F
+  }) => {
+    const signer = App.provider.getSigner()
+    console.log(signer)
```
Please remove the unneccessary logs

thongpt-codelink replied:
This is from the original. These logs also remain in the production version.

huynvk commented on src/static/js/earn.js:
```diff
@@ -73,6 +74,88 @@ async function main() {
+    const allowedTokens = await STAKING_TOKEN.allowance(App.YOUR_ADDRESS, STAKING_ADDR)
+    console.log(allowedTokens)
+    let allow = Promise.resolve()
+    if (allowedTokens / 1e18 == 0) {
```
Does `allowedTokens / 1e18` mean `allowedTokens` devides by 1e18? If this division equals to 0, does it mean allowedTokens is zero? Can we replace it by `allowedTokens == 0`?

thongpt-codelink replied:
It is from the original. I didn't notice the calculations.

huynvk commented on src/static/js/earn.js:
```diff
@@ -73,6 +74,88 @@ async function main() {
+    console.log(allowedTokens)
+    let allow = Promise.resolve()
+    if (allowedTokens / 1e18 == 0) {
+      snobMessage(`Approve spending`, `Please approve spending first. Please check your Metamask Wallet`, `information-circle-outline`, `primary`, false, `ok`);
```
Normally, we only use "\`" when we have some embedded variable inside (e.g. \`Hello, ${username}\`). In other cases, we only use single quote (') or double quote (") depend on team convention

thongpt-codelink replied:
It is from the original. They use this syntax for all functions to show messages.

huynvk commented on src/static/js/earn.js:
```diff
@@ -73,6 +74,88 @@ async function main() {
+    console.log(currentTokens)
+    const allowedTokens = await STAKING_TOKEN.allowance(App.YOUR_ADDRESS, STAKING_ADDR)
+    console.log(allowedTokens)
+    let allow = Promise.resolve()
```
What is the purpose of `allow`?

thongpt-codelink replied:
It is from the original. I see that the former dev like to create that "allow" promise for later promise chain syntax (when action submits)

huynvk commented on src/static/js/earn.js:
```diff
@@ -73,6 +74,88 @@ async function main() {
+    } else if (currentTokens / 1e18 > 0) {
+      halfmoon.toggleModal('modal-loading')
+      allow
+        .then(async function () {
```
You use arrow function in some places, and then `function ()` in other places. Can we make it consistent?

huynvk replied:
Can we make it consistent between using `promise.then` or `async / await`? Do you need `async` function here?

thongpt-codelink replied:
This snippet is from the original. The `async` is redundant but I didn't notice to remove it.

huynvk commented on src/static/js/earn.js:
```diff
@@ -73,6 +74,88 @@ async function main() {
+    let allow = Promise.resolve()
+    if (currentTokens / 1e18 > 0) {
+      halfmoon.toggleModal('modal-loading')
+      allow
```
Why do we need this `allow.then()`?

thongpt-codelink replied:
This is from the original. As the review above, former dev likes to create the "allow" promise then chain it

huynvk commented on src/static/js/earn.js:
```diff
@@ -1130,50 +1207,88 @@ async function main() {
+  function poolSNOB(options) {
+    //_print(``)
+    if (options.url) {
+      //_print(`<b>${options.pool_nickname}</b> <a href='${options.url}' target="_blank">${options.pool_name}</a>`)
```
Please remove unused code

thongpt-codelink replied:
This poolSNOB function is original.

huynvk commented on src/static/js/earn.js:
```diff
@@ -1130,50 +1207,88 @@ async function main() {
+      var combinedAprDisplay = '';
+      if (options.snowglobe_apr) {
+        let combinedAPR = options.icequeen_apr + options.snowglobe_apr
+        //_print(`Combined APR**: Day ${combinedAPR.toFixed(2)}% Week ${(combinedAPR * 7).toFixed(2)}% Year ${(combinedAPR * 365).toFixed(2)}%`)
```
Remove unused comment

huynvk commented on src/static/js/earn.js:
```diff
@@ -1130,50 +1207,88 @@ async function main() {
-    if ( options.staked_pool / 1e18 > 0 ) {
-      availableUnstake = stakeUnstake((options.staked_pool / 1e18).toFixed(6), false, 'S3F');
+    availableUnstake = '';
+    if ( options.staked_pool.amount / 1e18 > 0 ) {      
```
Can we replace this `options.staked_pool.amount / 1e18 > 0` by `options.staked_pool.amount > 0`

thongpt-codelink replied:
This condition is original. They come from different functions. The "green" is from poolSNOB(), the "red" is from poolS3F"

huynvk commented on src/static/js/earn.js:
```diff
@@ -1130,50 +1207,88 @@ async function main() {
-    if ( options.staked_pool / 1e18 > 0 ) {
-      availableUnstake = stakeUnstake((options.staked_pool / 1e18).toFixed(6), false, 'S3F');
+    availableUnstake = '';
+    if ( options.staked_pool.amount / 1e18 > 0 ) {      
```
Remove trailing spaces

huynvk commented on src/static/js/earn.js:
```diff
@@ -1183,26 +1298,25 @@ async function main() {
+      stakeBtn = `<button data-btn="${options.stake}" class="btn btn-sm mx-10 btn-success stakeBtn"><ion-icon name="lock-open-outline"></ion-icon> Stake PGL</button>`;
     }
-    if ( options.staked_pool / 1e18 > 0 ) {
+    if ( options.staked_pool.amount / 1e18 > 0 ) {
```
Same here with the condition

thongpt-codelink replied:
They come from different functions. The "green" is from poolSNOB(), the "red" is from poolS3F"

huynvk commented on src/static/js/earn.js:
```diff
@@ -1925,74 +1739,530 @@ async function main() {
+    </div>`
+  }
+  
+  function poolS3F(options){
```
This is a huge function. Is it that long originally? Can we split it smaller? That's OK to leave it as is if it was already that long before

thongpt-codelink replied:
It is long originally

Commit:
Code review: Format using prettier with the original config

Code review: Clean unused logs in stakingContract_stake

Code review: Remove unnecessary allow promise in stakingContract_stake

Code review: Change promise syntax to asyn/await for consistent in stakingContract_stake

Commit:
Code review: in stakingContract_withdraw:
- remove unused log
- remove unused AppSTAKING_CONTRACT parameter
- remove allow promise
- use async/await instead of promise chain
- use try catch to prevent duplicated snobMessage execution

Commit:
Code review: in stakingContract_stake:
- remove unnecessary division in the 1st condition

Commit:
Code review: in poolSNOB function:
- remove unused poolSNOB

Commit:
Fix new rendered Harvest SNOB unclickable
- change the wrong selector on unbind() function

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1,173 +1,1833 @@
+      stateMutability: 'nonpayable',
+      inputs: [
+        {type: 'address', name: '_token', internalType: 'address'},
+        {type: 'address', name: '_governance', internalType: 'address'},
```
These changes is due to the formatter.

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1,173 +1,1833 @@
+  const LINK_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x974Ef0bDA58C81F3094e124f530eF34fe70dc103'
+  const USDT_AVAX_TVL = 'https://info.pangolin.exchange/#/account/0x74dB28797957a52a28963F424dAF2B10226ba04C'
+
+  const stakingContract_stake = async ({
```
I cleaned the unused logs and use async/await to make it consistent

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -339,117 +2037,118 @@ async function main() {
-      res.pairs.forEach( p => {
-        if ( pairmatch(p, 'usdt', 'wavax') ) {
-          pool5tvl = p.locked;
+      res.pairs.forEach(p => {
```
These changes is due to the formatter

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -465,504 +2164,287 @@ async function main() {
 
-                <div onclick="toggleDetails('${options.pool_nickname}');" class="col-sm-12 col-md-1 align-items-center text-center text-md-right snob-tvl pb-10 pb-md-0 mx-auto">
-                    <ion-icon class="pointer" alt="More Details" name="chevron-down-outline"></ion-icon>
+  const aprDisplay = (cDayAPR, cWeekAPR, cYearAPR) => {
```
The aprDisplay function start from line **481** on the "red" side.

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -465,504 +2164,287 @@ async function main() {
+    </div>`
+  }
+
+  const calculateShare = (snowglobeContract, PAIR_ADDR, userSPGL, decimals, pool_percent) => {
```
The calculateShare function start from line **503** on the "red" column

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -465,504 +2164,287 @@ async function main() {
-      var eDayAPR = `${options.icequeen_apr.toFixed(2)}`;
-      var eWeekAPR = `${(options.icequeen_apr * 7).toFixed(2)}`;
-      var eYearAPR = `${(options.icequeen_apr * 365).toFixed(2)}`;
+  function pool(options) {
```
This is 2 different function, **pool** and ** poolS3D**. The pool function start from line **609** of the "red" column

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1016,197 +2500,241 @@ async function main() {
-  function poolS3F(options) {
-    var eDayAPR = options.icequeen_apr;
-    var eYearAPR = options.icequeen_apr * 365;
+  function poolS3D(options) {
```
This is 2 different functions too. poolS3D starts from line **871** of the red column

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1366,115 +2910,110 @@ async function main() {
-
-
   }
   function poolSNOB(options) {
```
I cleaned the unused codes

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1639,46 +3196,31 @@ async function main() {
-    pool_share_display: poolShareDisplay_8,
-    pool_share_display_pgl: '',
-    stake_display: ''
+  await renderPoolS3F({
```
I replaced the **poolS3F** using this function. This function is declared later using the old function declaration syntax to hoist.

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1860,140 +3414,625 @@ async function main() {
+    </div>`
+  }
+
+  function poolS3F(options) {
```
The **poolS3F** is an original function. It is long, but I don't know the business well enough to change.

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -1860,140 +3414,625 @@ async function main() {
+        </div>
+    </div>`
 
-  $(".claimBtn").click(function(){
```
This should compare with line **3454** on the "green" section

thongpt-codelink commented on src/static/js/earn.js:
```diff
@@ -2299,90 +4519,37 @@ const stakingContract_approve = async function (chefAbi, chefAddress, stakeToken
-      })
-  }
-}
-const stakingContract_stake = async function (chefAbi, chefAddress, stakeTokenAddr, App) {
```
The **stakingContract_stake** is moved to line **1611** in the "green" column
